### ISSUE TITLE
a confusion

### ISSUE THREAD

**Issue by Yinpenguin** [1991viet/Yolo-pytorch#4]
Hello,
       I have some trouble in reading your code.
       In the 'loss.py' file:
for i, anno in enumerate(ground_truth[b]):
                gt[i, 0] = (anno[0] + anno[2] / 2) / self.reduction    
                gt[i, 1] = (anno[1] + anno[3] / 2) / self.reduction     
                gt[i, 2] = anno[2] / self.reduction                    
                gt[i, 3] = anno[3] / self.reduction    
Thus, ground truth should be (xmin, ymin, width, height)                 
but in the 'voc_dataset.py', ground truth is (xmin, ymin, xmax, ymax)
I don't understand it, can you help me? @1991viet

**Comment by 1991viet** [comment]
Hi,
Thanks for the question. If you take a look at function Resize() at the end of this script https://github.com/1991viet/Yolo-pytorch/blob/master/src/data_augmentation.py, you will see that our "final" ground truth is (resized_xmin, resized_ymin, resize_width, resize_height, label). This function is the last step of the augmentation step, and it will always be done, either in training or test phase.
Bests,

**Comment by Yinpenguin** [comment]
Thanks for your patience with me, I was careless to read the ‘data_augmentation.py'.

**Comment by 1991viet** [comment]
You are welcome. I know that the code is long, hence there would be some parts we focus on less than the others.

**Comment by Yinpenguin** [comment]
Hi,
There are some errors when I run your code to train my model.  It looks like a version issue, and my version pytorch 0.4.1, but your code seems to be lower than 0.4.1. So, which pytorch version do you use  ?

**Comment by 1991viet** [comment]
Hi,
The code should be run well with pytorch ver >=0.4.0. Could you give me the error message?

**Comment by Yinpenguin** [comment]
BTW, when I run the 'test_voc_video.py', it worked  well.

**Comment by 1991viet** [comment]
Hi,
It seems weird, since in all train and test scripts, I use the same way to load pre-trained model. Could you try to run the code with pytorch 0.3.1? If you are familiar with creating virtual environment, then you could create one, and leave your default python interpreter intact :)
Another suggestion is that, you could even train my model **without** using any pretrained model (train from scratch). I tried once, and it still works, and converges. Of course it will take a bit longer time to converge.

**Comment by Yinpenguin** [comment]
Hi,
I tried it without pre-trained model just now, and it worked well.
One more question, it's about anchors and ground truth.
In your code, you align them on the top left, and computing IoU of them. Finally, you  choose best anchor of every ground truth . Besides
However, in the original papers， it seems different with you. Besides, the  formula to calculate target coord[0] and coord[1] is different to the paper.
Could you explain them to me ? Thank you.

**Comment by 1991viet** [comment]
Hi,
Could you point out on which part of my code (script and line(s)) you dont understand ?
Bests,

**Comment by Yinpenguin** [comment]
Hi,
There is a link describing my question in detail. (last question)
[http://note.youdao.com/noteshare?id=ed53063bf4ea71e7a1a6605177bd9367](url)
Best,

**Comment by 1991viet** [comment]
Hi,
Sorry now I am a bit busy with another object detection model. I will answer your question by Saturday, is that fine to you ? If it is urgent to you, just tell me
Thanks,

**Comment by Yinpenguin** [comment]
It's Ok, I'm in a hurry to solve this problem.

**Comment by 1991viet** [comment]
Hi,
I am sorry for the late response. Let me explain:
If you look at the script loss.py, lines 35, 36 and 44-47 and 56-59 are used for normalizing bbox prediction. You see, x and y are normalized to [0,1] by sigmoid,  then added by lin_x and lin_y respectively (lin_x and lin_y could be seen as "index" of cell (0-13)). The fomular are shown in page 3 on the paper https://arxiv.org/pdf/1612.08242.pdf 
The 2 lines you mentioned:
tcoord[b][best_n][0][gj * width + gi] = gt[i, 0] - gi
tcoord[b][best_n][1][gj * width + gi] = gt[i, 1] - gj
The loop contains these 2 lines is used to iteratively select the best anchor for each ground truth. Select here mean choosing the best one, but do not adapt (make it fit better to gt).
Hope that my explanation could somehow help you.
Bests,

**Comment by Yinpenguin** [comment]
Hi,
Many thanks to your explanation, now I understand.
Bests,

**Comment by tutay** [comment]
Hi viet,
I need to ask you about something. In your loss function line 140-141. 
why did you divide the ground truth w,h to best selected anchor width and height and take the logarithm.

`tcoord[b][best_n][2][gj * width + gi] = math.log(max(gt[i, 2], 1.0) / self.anchors[best_n, 0])
                tcoord[b][best_n][3][gj * width + gi] = math.log(max(gt[i, 3], 1.0) / self.anchors[best_n, 1])`

I cannot see this in the original paper. Most probably i missed something. Thanks in advance.
